docs/design-docs/08-package-variant.md

```diff
@@ -0,0 +1,234 @@
+advanced use cases, we may need to specialize the package based upon other
+contextual information. This particularly comes into play when the user
+deploying the workload does not have direct control over the context in which
+it is being deployed. For example, one part of the organization may manage the
+infrastructure - that is, the cluster in which we are deploying the workload -
+and another part the actual workload. We would like to be able to pull in inputs
+specified by the infrastructure team automatically, based the cluster to which
+we are deploying the workload, or perhaps the region in which that cluster is
+deployed.
+
+To facilitate this, the package variant controller can "inject" configuration
+directly into the package. This means it will use information specific to this
+instance of the package to lookup a resource in the Porch cluster, and copy that
+information into the package. Of course, the package has to be ready to receive
+this information. So, there is a protocol for facilitating this dance:
+- Packages may contain resources annotated with `kpt.dev/config-injection`
+- Usually, these will also be `config.kubernetes.io/local-config` resources, as
+  they are likely just used by local functions as input. But this is not
+  mandatory.
+- The package variant controller will look for any resource in the Kubernetes
+  cluster matching the Group, Version, and Kind of the package resource, and
+  satisfying the *injection selector*.
+- The package variant controller will copy the `spec` field from the matching
+  in-cluster resource to the in-package resource.
+
+Note that because we are injecting data *from the Kubernetes cluster*, we can
+also monitor that data for changes. For each resource we inject, the package
+variant controller will establish a Kubernetes "watch" on the resource. A change
+to that resource will result in a new Draft package with the updated
+configuration injected.
+
+[Note: it may be necessary for the package variant controller to annotate the
+in-package resource with a hash of the in-cluster resource to detect changes,
+this should be discussed].
+
+### Namespace Configuration and Injection
+[Note: not yet implemented]
+Creating a namespace and/or setting the namespace for a particular package is a
+very common operation. However, since namespace provisioning in a cluster is a
+privileged operation, the deployer of a package may not have the authority to
+provision a namespace. For this reason, upstream packages should not directly
+include `Namespace` resources if they want to be truly reusable.
+
+So, the package variant controller provides some convenience features for
+targeting the package at a particular namespace. First, it is possible to use
+the KRM function capability to call the well-known `set-namespace` function.
+For convenience, you can alternatively set the `namespace.value` field in the
+`PackageVariant` resource, and this will be done for you automatically.
```

Comment:
I think with the KRM functions capability, adding the `namespace.value` field in the `PackageVariant` resource is not necessary. I feel it is best to have a singular way to achieve this task (to avoid user confusion), and the `set-namespace` function seems convenient enough.

Reply:
`set-namespace` doesn't have a "Create" option, though.

Reply:
Probably the best thing to do would be as you suggest, but add a `create` option to `set-namespace`, or have the user use `upsert-resource`, though the UX on that would be painful.

Reply:
Updated to reflect this.